Initial ObjectFormat implementation

0xdaryl opened the pull request:
* Initial ObjectFormat and FunctionCallData definitions
* Create X86 specialization of FunctionCallData
* Initial AMD64 implementation of JitCodeRWXObjectFormat
* Initial AMD64 implementation of JitCodeRXObjectFormat

Issue: #5569

DanHeidinga commented on compiler/x/CMakeLists.txt:
```diff
@@ -68,6 +68,8 @@ if(TR_TARGET_BITS STREQUAL 64)
 		${CMAKE_CURRENT_LIST_DIR}/amd64/codegen/AMD64SystemLinkage.cpp
 		${CMAKE_CURRENT_LIST_DIR}/amd64/codegen/OMRELFRelocationResolver.cpp
+                ${CMAKE_CURRENT_LIST_DIR}/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp
+                ${CMAKE_CURRENT_LIST_DIR}/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp
```
nitpick - tabs vs spaces

0xdaryl replied:
Fixed.  Thanks.

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
@r30shah @fjeremic : I would appreciate a review from you please.

This PR simply introduces the API and an implementation on AMD64 for the RWX and RX object formats.  It does not provide a 32-bit object format (that will be coming later).  I implemented it slightly differently from the design in #5569 in that it only provides an API for a generic "function call" rather than separate APIs for "global functions" and "code cache functions". I found there was too much ambiguity in deciding when to choose one over the other that I went with a single API instead.

This work needs to be introduced in stages to prevent downstream build breakages.  The next PR after this one in OMR will use the API on AMD64 defined in this PR.

0xdaryl commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+      }
+
+   ccFunctionData *ccFunctionDataAddress =
+      reinterpret_cast<ccFunctionData *>(data.cg->allocateCodeMemory(sizeof(ccFunctionData), false));
```
**Note to reviewers:**  this implementation uses `allocateCodeMemory`.   This is only a temporary solution until #5634 is merged, in which case it will switch to allocating such memory from CCData.

0xdaryl commented:
Interesting.  The macOS porttest failed despite the fix in #5632 being included.

fjeremic commented on compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:
```diff
@@ -0,0 +1,217 @@
+
+   data.cg->resetIsLeafMethod();
+
+   TR::Instruction *callInstr;
```
We should NULL initialize this.

fjeremic commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+   uint8_t modRM = data.isCall ? 0x15 : 0x25;  // RIP addressing mode
+   *data.bufferAddress++ = modRM;
+
+   TR_ASSERT_FATAL(comp->target().cpu.isTargetWithinRIPRange(
+                      reinterpret_cast<intptr_t>(ccFunctionDataAddress),
+                      reinterpret_cast<intptr_t>(data.bufferAddress+4)),
+                      "ccFunctionData must be reachable directly");
```
This assert should be improved to print the values of the two arguments. We should also use the `_WITH_NODE` assert here as well.

fjeremic commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+                      "ccFunctionData must be reachable directly");
+
+   int32_t disp32 = (int32_t)(reinterpret_cast<intptr_t>(ccFunctionDataAddress) - reinterpret_cast<intptr_t>(data.bufferAddress+4));
+   *(int32_t *)data.bufferAddress = disp32;
```
`reinterpret_cast` need here as well as elsewhere. Will stop commenting on this one.

fjeremic commented on compiler/x/objectfmt/FunctionCallData.hpp:
```diff
@@ -0,0 +1,224 @@
+    * @param[in] parm_cg : \c TR::CodeGenerator object
+    * @param[in] parm_isCall : if true, emit a call instruction; if false, emit a jump instruction
+    */
+   FunctionCallData(
+         TR::SymbolReference *parm_methodSymRef,
+         TR::Node *parm_callNode,
+         TR::Register *parm_scratchReg,
+         uintptr_t parm_targetAddress,
+         TR::RegisterDependencyConditions *parm_regDeps,
+         int32_t parm_reloKind,
+         bool parm_useSymInstruction,
+         int32_t parm_adjustsFramePointerBy,
+         TR::CodeGenerator *parm_cg,
+         bool parm_isCall = true) :
```
I guess this can be subjective, but to me the `TR::CodeGenerator*` is _always_ needed for these objects to be constructed properly in each overload of this constructor. Where as the other parameters may change. To me it makes most sense to put all the parameters which are consistently _always_ present first. Given these APIs are exclusively used in the codegen, we should just put the `TR::CodeGenerator*` parameter first.

r30shah commented on compiler/objectfmt/FunctionCallData.hpp:
```diff
@@ -0,0 +1,52 @@
+   {
+public:
+
+   FunctionCallData(
```
Is the FunctionCallData needed to be short lived object only? Do we need `TR_ALLOC` for the FunctionCallData in case we want to cache the information? One example, I was thinking and currently trying out is in Snippet API where as I mentioned in https://github.com/eclipse/omr/pull/5575/files#r517395961, We can add `print` API and for that if we have cached the function call data to snippet, it will be easier.

dsouzai replied:
> Do we need TR_ALLOC for the FunctionCallData in case we want to cache the information?

All `TR_ALLOC` does is add a bunch of `placement new` definitions. You don't technically need it since you _could_ just do something like
```
void *storage = allocateHeapMemory(sizeof(FunctionCallData));
FunctionCallData *fcData = new (storage) FunctionCallData(...);
```
Obviously this is terrible code, but the point being that `TR_ALLOC` doesn't give you any new caching ability, just that it allows cleaner code, ie
```
FunctionCallData *fcData = new (comp->trHeapMemory()) FunctionCallData(...);
```

r30shah replied:
Okk yes, We do not need `TR_ALLOC` for allocating object on the heap. So let me rephrase my question, can we add it to list of object types in TR_Memory so that it can be used in more cleaner code as mentioned by @dsouzai  in previous comment?

ymanton commented on compiler/codegen/OMRCodeGenerator.hpp:
```diff
@@ -1445,6 +1446,27 @@ class OMR_EXTENSIBLE CodeGenerator
+    */
+   void createObjectFormat() { return; }
+
+   /**
+    * @return TR::ObjectFormat created for this \c TR::CodeGenerator.  If the value
+    *         returned is NULL then ObjectFormats are not supported.
+    */
+   TR::ObjectFormat *getObjFmt() { return _objectFormat; }
```
Agree. Let's just stick with ObjectFormat, short names are hard to guess if you don't already know they exist.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:
```diff
@@ -0,0 +1,217 @@
+            }
+
+         TR_X86OpCodes op = data.isCall ? CALLReg : JMPReg;
+         if (data.prevInstr)
+            {
+            callInstr = generateRegInstruction(data.out_loadInstr, op, data.scratchReg, data.regDeps, data.cg);
+            }
```
Perhaps I'm not familiar enough with this area of the code, but this looks a bit unintuitive to me. Every other occurrence of this pattern passes `data.pervInstr` to `generate*Instruction`, but here we pass `data.out_loadInstr`. Maybe a comment explaining the reason would help?

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:
```diff
@@ -0,0 +1,217 @@
+uint8_t *
+OMR::X86::AMD64::JitCodeRWXObjectFormat::encodeFunctionCall(TR::FunctionCallData &data)
+   {
+   uint8_t op = data.isCall ? 0xe8 : 0xe9;   // CALL rel32 | JMP rel32
```
This can be a const.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:
```diff
@@ -0,0 +1,217 @@
+OMR::X86::AMD64::JitCodeRWXObjectFormat::encodeFunctionCall(TR::FunctionCallData &data)
+   {
+   uint8_t op = data.isCall ? 0xe8 : 0xe9;   // CALL rel32 | JMP rel32
+   *data.bufferAddress++ = op;
```
Personally, I find this kind of combined dereference and increment a bit difficult to read.

Leonardo2718 replied:
Fair enough.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:
```diff
@@ -0,0 +1,217 @@
+   uint8_t op = data.isCall ? 0xe8 : 0xe9;   // CALL rel32 | JMP rel32
+   *data.bufferAddress++ = op;
+
+   int32_t disp32 = data.cg->branchDisplacementToHelperOrTrampoline(data.bufferAddress+4, data.methodSymRef);
```
This too can be a const.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp:
```diff
@@ -0,0 +1,217 @@
+   *data.bufferAddress++ = op;
+
+   int32_t disp32 = data.cg->branchDisplacementToHelperOrTrampoline(data.bufferAddress+4, data.methodSymRef);
+   *(int32_t *)data.bufferAddress = disp32;
```
This should b a reinterpret_cast.

Leonardo2718 commented on compiler/objectfmt/JitCodeRXObjectFormat.hpp:
```diff
@@ -0,0 +1,41 @@
+   {
+public:
+
+   JitCodeRXObjectFormat() : OMR::JitCodeRXObjectFormatConnector() {}
```
I don't think this constructor declaration is needed.

0xdaryl commented:
I simply rebased and pushed to resolve the merge conflict.  I am working on addressing the comments now.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+TR::Instruction *
+OMR::X86::AMD64::JitCodeRXObjectFormat::emitFunctionCall(TR::FunctionCallData &data)
+   {
+   TR::SymbolReference *methodSymRef;
```
Maybe this should be NULL initialized to be safe.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+   if (comp->isRecursiveMethodTarget(methodSymRef->getSymbol()) &&
+       !comp->couldBeRecompiled())
+      {
+      TR::X86ImmSymInstruction *callImmSymInstr;
```
Maybe this should be NULL initialized to be safe.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+   functionDataSymbol->setNotDataAddress();
+   TR::SymbolReference *functionDataSymRef = new (comp->trHeapMemory()) TR::SymbolReference(comp->getSymRefTab(), functionDataSymbol, 0);
+
+   TR::X86CallMemInstruction *callInstr;
```
Maybe this should be NULL initialized to be safe.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+uint8_t *
+OMR::X86::AMD64::JitCodeRXObjectFormat::encodeFunctionCall(TR::FunctionCallData &data)
+   {
+   TR::SymbolReference *methodSymRef;
```
Maybe this should be NULL initialized to be safe.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+   ccFunctionDataAddress->address = targetAddress;
+
+   *data.bufferAddress++ = 0xff;  // CALL | JMP Mem
+   uint8_t modRM = data.isCall ? 0x15 : 0x25;  // RIP addressing mode
```
This can be a const.

Leonardo2718 commented on compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp:
```diff
@@ -0,0 +1,187 @@
+                      reinterpret_cast<intptr_t>(data.bufferAddress+4)),
+                      "ccFunctionData must be reachable directly");
+
+   int32_t disp32 = (int32_t)(reinterpret_cast<intptr_t>(ccFunctionDataAddress) - reinterpret_cast<intptr_t>(data.bufferAddress+4));
```
I think this can be a const.

0xdaryl commented:
@r30shah @fjeremic @Leonardo2718 : Thanks for your careful reviews and sorry for the delay in responding.  I believe I have addressed all the comments that were raised.  I would appreciate a second review to ensure your comments or questions were addressed.  Thank you!

fjeremic commented on compiler/objectfmt/OMRFunctionCallData.hpp:
```diff
@@ -0,0 +1,118 @@
+    * the address where that function address is written.
+    */
+   uint8_t *out_encodedMethodAddressLocation;
+
```
We should likely make this a `struct` then? Minor detail I suppose. `struct` fields are public by default where as `class` fields are private by default.

0xdaryl replied:
Yes, you're right.  But I left it as a class because it is decorated `OMR_EXTENSIBLE`.  I didn't think through all the implications of making "extensible structs".  In theory we shouldn't have a problem (maybe some work on the linter?) but perhaps something to be solved in a different PR.

fjeremic reviewed (approved):
Looks reasonable to me with all the new updates addressing previous concerns.

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
If everyone is ok with this, could a committer please merge?

For additional integration testing, I have tested it builds successfully on all platforms with the latest OpenJ9 master.

Leonardo2718 commented on compiler/x/objectfmt/FunctionCallData.hpp:
```diff
@@ -0,0 +1,224 @@
+
+   /**
+    * APIs for non-helper function calls
+    */
```
Ok, that's fine.

My main concern here is that making sure the correct overload is picked can be error-prone, so having a descriptive name would disambiguate which version the programmer intended to call. Also, it's unclear right now what invariants each constructor is enforcing.

Leonardo2718 commented:
LGTM :+1: 

@fjeremic you're the assigned committer on this one :wink:
